Consistently use CFLAGS / LIBS in Makefiles

chu11 opened the pull request:
Per #3776, I found that ZMQ_CFLAGS and ZMQ_LIBS were used excessively throughout the Makefiles.  Libraries /modules that have no need to link to zmq (e.g. `liboptparse`) were linked to it.  Some of these were likely old additions that were no longer needed after refactoring with `libczmqcontainers`.   Others were likely lazy cut & paste issues.   As I went through, realized that many ZMQ_LIBS additions were unnecessary b/c other libs (e.g. `libflux-internal`) already links to it.  So cleaned those up too.

Along the way, noticed more unnecessary inclusions and just decided to go nuts, cleaning up LIBUUID, LIBSODIUM, HWLOC_CFLAGS, LIBDL, LIBRT, and JANSSON macros as well.  Some cases were stupid, where variables like `test_ldadd` included `FOO_LIBS`, and every test binary that used `test_ldadd` also linked to `FOO_LIBS`.

I may have missed some, but minimally its much more cleaned up.

garlick commented:
As noted in #3617, a _module_ picks up all of the objects in a libtool convenience library in its LIBADD, and thus requires all external dependencies of the convenience library.  However, an _executable_ only pulls in objects that resolve symbols in the executable, and would only require the external dependencies of those objects.

If we go with this cleanup, then our commands will possibly end up linked against external libraries they don't use.

OTOH if we go the other way and drop external libraries from the convenience library's LIBADD, those eight(?) libraries will have to be added to every module's LIBADD, and a subset to every executable's LDADD.  A lot of repetition.  Grr.

Maybe we could go with this cleanup, then see if we can reduce libflux-internal's external dependencies by breaking out a few key bits?   Obviously the key one is going to be `$(ZMQ_LIBS)`.  We may be able to get rid of `$(LIBSODIUM_LIBS)` after #3785 gets merged.   Maybe `$(LIBUUID_LIBS)`.

garlick commented:
Just did an experiment: build master and run
```
$ cd src/cmd/.libs
$ for i in *; do echo -n "$i "; ldd $i|wc -l; done
flux 33
flux-event 30
flux-exec 30
flux-job 30
flux-keygen 27
flux-kvs 30
flux-logger 29
flux-module 30
flux-ping 30
flux-queue 30
flux-R 11
flux-start 30
flux-terminus 30
```
I then repeated that for this branch, and got the same results.  So this PR does no harm relative to current master.

chu11 commented:
> Maybe we could go with this cleanup, then see if we can reduce libflux-internal's external dependencies by breaking out a few key bits? Obviously the key one is going to be $(ZMQ_LIBS). We may be able to get rid of $(LIBSODIUM_LIBS) after #3785 gets merged. Maybe $(LIBUUID_LIBS).

This is sort of my feeling, it's step one on the path to something better.  Many of the CFLAGS/LIBS were cut & paste errors or forgotten removals after refactoring.  It made it hard to break out ZMQ_LIBS b/c *everything* was linking to it.

garlick commented on src/connectors/shmem/Makefile.am:
```diff
@@ -23,5 +23,4 @@ shmem_la_LDFLAGS = -module $(san_ld_zdef_flag) \
 shmem_la_LIBADD = \
 	$(top_builddir)/src/common/libflux-internal.la \
 	$(top_builddir)/src/common/libflux-core.la \
-	$(ZMQ_LIBS) \
```
shmem connector uses the `zsock` class, so this is still required.  Also it looks like `$(ZMQ_CFLAGS)` is missing so that should be added.

garlick commented on src/test/Makefile.am:
```diff
@@ -8,8 +8,7 @@ AM_LDFLAGS = \
 	-I$(top_srcdir)/src/include \
-	-I$(top_builddir)/src/common/libflux \
-	$(ZMQ_CFLAGS)
+	-I$(top_builddir)/src/common/libflux
```
There are no more c targets in src/test, so AM_CPPFLAGS, AM_LDFLAGS, AM_CFLAGS, and LDADD can all go away.

garlick commented on src/common/libjob/Makefile.am:
```diff
@@ -56,8 +56,7 @@ test_ldadd = \
 	$(LIBRT) \
         $(FLUX_SECURITY_LIBS) \
-	$(LIBDL) \
-	$(LIBSODIUM_LIBS)
```
libjob/sign_none.c uses libsodium, so this should remain.

Another option would be to convert it to use the CCAN base64 just merged.

chu11 replied:
working on #3787, perhaps that should go in first before this

garlick commented on src/cmd/Makefile.am:
```diff
@@ -21,7 +21,6 @@ fluxcmd_ldadd = \
 	$(top_builddir)/src/common/libflux-optparse.la \
 	$(FLUX_SECURITY_LIBS) \
 	$(LIBPTHREAD) \
-	$(LIBDL) \
```
flux-module.c includes dlcfn.h but doesn't appear to use it.  Maybe drop that too?

garlick commented on t/Makefile.am:
```diff
@@ -454,204 +451,167 @@ mpi_mpich_basic_simple_LDADD = $(MPI_CLDFLAGS) $(test_ldadd)
 module_parent_la_SOURCES = module/parent.c
 module_parent_la_CPPFLAGS = $(test_cppflags)
 module_parent_la_LDFLAGS = $(fluxmod_ldflags) -module -rpath /nowher
-module_parent_la_LIBADD = \
-	$(test_ldadd) $(LIBDL)
```
t/module/parent.c calls `dlopen()` so this should remain.

garlick commented on src/common/libjob/Makefile.am:
```diff
@@ -54,7 +54,6 @@ test_ldadd = \
         $(top_builddir)/src/common/libtap/libtap.la \
```
Commit message:   "macro was more than necessary'.  Missing "used"?

That was commit 6234eb20fae7ced9ada4c22ddf76a7d6230c7afa

garlick commented on src/common/libjob/Makefile.am:
```diff
@@ -52,7 +52,6 @@ test_ldadd = \
         $(top_builddir)/src/common/libflux/libflux.la \
         $(top_builddir)/src/common/libflux-internal.la \
         $(top_builddir)/src/common/libtap/libtap.la \
-	$(JANSSON_LIBS) \
```
I think we need this one.  jansson is used in several .c files in libjob.

garlick commented on src/common/libkvs/Makefile.am:
```diff
@@ -61,7 +61,6 @@ test_ldadd = \
 	$(top_builddir)/src/common/libflux/libflux.la \
 	$(top_builddir)/src/common/libflux-internal.la \
 	$(top_builddir)/src/common/libtap/libtap.la \
-	$(JANSSON_LIBS) \
```
jansson is used in libkvs/kvs_commit.c so presumably we should keep this and add missing `$(JANSSON_CFLAGS)`.

The jansson header is included in kvs.c, but I think it could be removed.

garlick commented on src/common/librlist/Makefile.am:
```diff
@@ -26,7 +26,6 @@ test_ldadd = \
 	$(top_builddir)/src/common/libtap/libtap.la \
 	$(top_builddir)/src/common/libflux-internal.la \
 	$(LIBPTHREAD) \
-	$(JANSSON_LIBS) \
```
librlist uses jansson extensively.  Should add the missing `$(JANSSON_CFLAGS)`.

garlick commented on src/modules/job-exec/Makefile.am:
```diff
@@ -58,7 +58,7 @@ test_ldadd = \
 	$(top_builddir)/src/common/libflux-internal.la \
 	$(top_builddir)/src/common/libflux-core.la \
-	$(LIBPTHREAD) $(JANSSON_LIBS)
+	$(LIBPTHREAD)
```
job-exec uses jansson directly, but is missing a `$(JANSSON_LIBS)` in its LIBADD.

garlick commented on src/modules/job-manager/Makefile.am:
```diff
@@ -112,7 +112,7 @@ test_ldadd = \
 	$(top_builddir)/src/common/libjob/libjob.la \
 	$(top_builddir)/src/common/libflux-internal.la \
 	$(top_builddir)/src/common/libflux-core.la \
-	$(LIBPTHREAD) $(JANSSON_LIBS)
```
Same here, $(JANSSON_LIBS) Is missing from LIBADD.

Also, what's the deal with `$(fluxmod_libadd)`?  All the modules add it, but it doesn't seem to be defined anywhere?  (possibly I'm missing something).

chu11 replied:
> Also, what's the deal with $(fluxmod_libadd)? All the modules add it, but it doesn't seem to be defined anywhere? (possibly I'm missing something).

I'm going to take a guess.  Perhaps `fluxmod_ldflags` was slowly copied around and cutters/pasters assumed `fluxmod_libadd` or `fluxmod_ldadd` was also defined?  Apparently this issue has existed as far back as `0.1.0`,  I'll do a bit of cleanup with this too.

garlick reviewed (changes requested):
I found a few places where I didn't think removal was appropriate - where a module/command/whatever was using the dependency directly, as opposed to needing it because it linked with a convenience library that uses it.

chu11 commented:
> Edit: by the way I'm assuming that if a library is used directly, it should have both the CFLAGS and the LIBS def for it in its makefile, even if it uses a convenience library that also uses that library. In that case the LIBS might be redundant, but at least if the convenience library stops using the library, other things won't need to be updated.

So I did the opposite.  B/c `libflux-core` and `libflux-internal` effectively link to everything, and a very large swath of stuff always ldadd `libflux-core` and `libflux-internal`, the redundancy sort of existed everywhere.  Which is why I may have removed it in more places than you originally thought I should have.  And (unsurprisingly) we're not consistent.  There are locations that we should probably add ZMQ_LIBS (or some other libs) for consistency.

Still thinks we should add the XXX_LIBS back?

> The CFLAGS is needed regardless, so the include file can be found if it was installed to a non-standard prefix.

Agreed, I (atleast intended) to do this, but could have made a mistake or two (i looks like I did given above).

garlick commented:
> Still thinks we should add the XXX_LIBS back when a lib/module/executable uses it?

If we need the CFLAGS anyway, and the libs that the convenience libs depend on may be changing, then it seems to me to make the most sense to add/keep both where the dependency is used directly.

chu11 commented:
Another comment.  Lots of ldadds (especially test_ldadds) are through `common/libflux-core.la`.  This is different than using `common/libflux/libflux.la`, the latter which does not yet link to ZMQ_LIBS and a bunch of other libs.

Another approach is to convert internal ldadds to always use the internal libs (e.g. `librouter/librouter.la`) instead of the wrapper that already does all the linking (`libflux-internal.la`).

It would force us to put ZMQ_LIBS, LIBUUID_LIBS, etc. into the right places.

chu11 commented:
> If we need the CFLAGS anyway, and the libs that the convenience libs depend on may be changing, then it seems to me to make the most sense to add/keep both where the dependency is used directly.

Ok, I'll make another run through it, since this approach will be different.

garlick commented:
> Lots of ldadds (especially test_ldadds) are through common/libflux-core.la. This is different than using common/libflux/libflux.la

I think it's good to link against `libflux-core.la` where needed.  The only places where there may be an issue is in subdirectories of `src/common`, when iteratively working on something that becomes part of `libflux-core.la`.  You want the test to use the latest version of the component under development, but there is no makefile rule to rebuild `libflux-core.la` at that level.  There IMHO it makes sense to reference the component directly.

Are there other reasons to go the other way?  (I'm not thinking of them right now)

chu11 commented:
> Are there other reasons to go the other way? (I'm not thinking of them right now)

Its just to reduce redundancy of linking, and to some extent force us to only link what we need to link.

garlick commented:
> Its just to reduce redundancy of linking, and to some extent force us to only link what we need to link.

If it decreases the `make check` time I'm all for it.  Otherwise, meh?  (My opinion, and not a cherished one :-)

chu11 commented:
re-pushed addressing the comments / discussion above.

- 3 random cleanup commits before the beefier stuff
- generally use the rule to add/use CFLAGS / LIBS when code uses stuff from that lib.  So in addition to the removals I had before, there are now commits to add CFLAGS / LIBS in places.
- adjust commit messages as a result

garlick commented on src/broker/Makefile.am:
```diff
@@ -93,14 +95,16 @@ test_ldadd = \
 
 test_cppflags = \
         -I$(top_srcdir)/src/common/libtap \
-        $(AM_CPPFLAGS)
+        $(AM_CPPFLAGS) \
+	$(JANSSON_CFLAGS)
```
Inconsistent whitespace here (use tabs I think).

Also `$(JANSSON_CFLAGS)` is also being added to `$(AM_CPPFLAGS)`, which is included here, so this is a dup.

garlick commented on src/modules/job-list/Makefile.am:
```diff
@@ -9,7 +9,8 @@ AM_CPPFLAGS = \
 	-I$(top_srcdir)/src/include \
 	-I$(top_builddir)/src/common/libflux \
-	$(FLUX_SECURITY_CFLAGS) $(YAMLCPP_CFLAGS)
+	$(FLUX_SECURITY_CFLAGS) $(YAMLCPP_CFLAGS) \
```
What's YAMLCPP_CFLAGS doing here?  Leftover from days gone by?  It's not defined.  Maybe drop that from here and other places it was cut and paste (job-archive, job-info, job-ingest, and kvs-watch).

chu11 replied:
oh, haha!  I just assumed we were using it :P  i'll yoink it

garlick reviewed (approved):
Just spotted a couple minor things but generally looking good!

chu11 commented:
re-pushed per comments above, there's one extra commit in there to deal w/ the YAMLCPP_CFLAGS.  @garlick wanna take a quick skim before I label MWP.

garlick reviewed (approved):
Looks good after the small update.  Thanks!
